feat(merge): allow to always append pr to queue (#1097)
Mergifyio/mergify-engine: doc/source/actions.rst

         against its base branch at a time.
         This allows you to e.g., save CI time, as Mergify will queue the
         mergeable pull requests and update them serially, one at a time.

>          But if the pull request is ready and already in sync with its base
>          branch, the queue is bypassed and the pull request is directly merged

Reviewer: ```suggestion
         If the pull request is already in sync with its base branch, the queue is
         bypassed and the pull request is directly merged.
```

>        * ``smart-ordered`` enables :ref:`strict merge`, same behavior as before

Reviewer: ```suggestion
       * ``smart+ordered`` enables :ref:`strict merge`, same behavior as before
```
I'd prefer we use the `+` symbol instead.

>        * ``smart-ordered`` enables :ref:`strict merge`, same behavior as before
>          except queue is never bypassed, pull requests are always added at the
>          end of the queue

Reviewer: ```suggestion
       * ``smart-ordered`` enables :ref:`strict merge` with the same behavior as ``smart``,
         except that the queue is never bypassed. Pull requests are always added to the
         end of the queue, even if they are ready to be merged.
```

       Note that the ``rebase`` method has many drawbacks, see :ref:`strict
       rebase`.

   * - ``strict_smart_behavior``
     - string
     - ``merge-if-ready``
     - Behavior when a pull request is ready to be merge in ``strict=smart``
       mode. Possible values:

       * ``merge-if-ready`` if the pull request is ready and already in sync
         with its base branch, the queue is by-passed and the pull request is

>          directly merged. This behavior allows to safe CIs resources.

Author: If you have a long-running CIs, you really save resources.

Assume CI takes 2 hours:

merge-if-ready:
* PR 1 is ready except CIs CI already ran since 1 hour
* PR 2 just become ready and already up2date, we just merge it without queuing it.
* PR 1 get rebased, previous CI RUN get canceled, we wait 2h and  merged it

CIs total runs times 3 hours

always-append:
* PR 1 is ready except CIs, CI already run since 1 hour
* PR 2 just become ready and already up2date, we put it in queue.
* PR 1 CIs finish 1 hour later and get merged
* PR 2 is rebased, CIs rerun and take 2 hours
* PR 2 get merged

CIs total runs times: 4 hours

>          directly merged. This behavior allows to safe CIs resources.

Reviewer: I think you forgot some:

`merge-if-ready`:

- PR 1 is ready except CIs CI already ran since 1 hour (**1 hour CI**)
- PR 2 just become ready and already up2date, we just merge it without queuing it (**2 hours CI**)
- PR 1 get rebased, previous CI RUN get canceled, we wait 2h and merged it (**2 hours CI**)

CIs total runs times **5 hours**

`always-append`:

- PR 1 is ready except CIs, CI already run since 1 hour (**1 hour CI**)
- PR 2 just become ready and already up2date, we put it in queue (**1 hours CI** since it's going to be cancelled)
- PR 1 CIs finish 1 hour later and get merged (**1 hour CI**)
- PR 2 is rebased, CIs rerun and take 2 hours (**2 hours CI**)
- PR 2 get merged

CIs total runs times: **5 hours**

So it's the same, no?

>          directly merged. This behavior allows to safe CIs resources.

Author: > I think you forgot some:
> 
> `merge-if-ready`:
> 
>     * PR 1 is ready except CIs CI already ran since 1 hour (**1 hour CI**)
> 
>     * PR 2 just become ready and already up2date, we just merge it without queuing it (**2 hours CI**)

PR 2 here is already rebased, CIs already green, so we do not have this 2 hours

>     * PR 1 get rebased, previous CI RUN get canceled, we wait 2h and merged it (**2 hours CI**)
> 
> 
> CIs total runs times **5 hours**
> 
> `always-append`:
> 
>     * PR 1 is ready except CIs, CI already run since 1 hour (**1 hour CI**)
> 
>     * PR 2 just become ready and already up2date, we put it in queue (**1 hours CI** since it's going to be cancelled)
>     * PR 1 CIs finish 1 hour later and get merged (**1 hour CI**)
> 
>     * PR 2 is rebased, CIs rerun and take 2 hours (**2 hours CI**)
> 
>     * PR 2 get merged
> 
> 
> CIs total runs times: **5 hours**
> 
> So it's the same, no?

Another way to explain it:

In always-append, things are always rebased (except if the queue is empty).
So for a PR, CIs run 2 hours before "merge" action is triggered and then 2 hours after the rebase. For 2 PRs, we have 8 hours

In merge-if-ready, if the PR doesn't need to be rebased, we merge it directly whatever the queue status is and skip the second 2 hours CIs (We still have 2 hours before "merge" action is triggered).
But the counter-part is that the first PR of the queue get its CIs canceled after 1 hour of run, rebased again and then 2 more hours of CIs , so 5 hours in total (2 hours before "merge" action is triggered, 1 hour for nothing, 2 hours after the rebased)
Here, for the two PRs, we have 7 hours in total.

>          directly merged. This behavior allows to safe CIs resources.

Reviewer: > PR 2 here is already rebased, CIs already green, so we do not have this 2 hours

Of course, you have: for the CI to be green, it has to run for 2 hours. :)

> In always-append, things are always rebased (except if the queue is empty).
So for a PR, CIs run 2 hours before "merge" action is triggered and then 2 hours after the rebase. For 2 PRs, we have 8 hours

👍  that's a good summary. In `always-append`, you always have 2 CI runs; the _election_ CI runs, where you make sure it works and it can enter the queue, and the _merge_ CI run that is done once it's at the top of the queue and going to be merged.

> In merge-if-ready, if the PR doesn't need to be rebased, we merge it directly whatever the queue status is and skip the second 2 hours CIs (We still have 2 hours before "merge" action is triggered).

The thing is that the more you cancel, the more CI time you use for nothing.
In `always-append` the CI time is constant and always _election_+_merge_. In `merge-if-ready`, you always have _election_, but then your _merge_ can be between 0 (the _election_ run is used) or infinite (you keep being canceled).

So it *can* save CI time, but it can also consume CI time forever 😨 😆

       * ``always-append`` pull requests are always placed at the end of the queue

> 

Author: good idea !

   * - ``strict_smart_priority_labels``
     - list of string
     - `[]`
     - List of labels to prioritize pull request in queue. Labels must be
       ordered from the highest priority to the lowest.

       Once a pull request enters in the queue, the priority can be changed even

>        if the label is removed.

Reviewer: I'm wondering if we shouldn't do something different, which is by using `conditions`. Labels are good but they are kind limited.

For example you might want PR from a bot to be always low priority. You could do:

```yaml
- name: automatic merge from dependabot
    conditions:
      - author~=^dependabot(|-preview)\[bot\]$
      - "status-success=ci/circleci: py37"
    actions:
      merge:
        strict: "smart"
        priority: -10
        method: rebase
  - name: automatic merge
    conditions:
      - base=master
      - "status-success=ci/circleci: py38"
      - "#approved-reviews-by>=1"
    actions:
      merge:
        strict: "smart"
        method: rebase
        priority: 10
```

I don't like priority being numbered because you never know which is low which is high.
So I'd say make them an `Integer` where the higher the more priority, so it's flexible, plus add some conveninent alias such as:
- `high` = 128
- `normal` = 0
- ` low` = -128

It makes it easy to understand when writing/reading the configuration.

Author: I have thought of this solution too (I have already a branch with that), I prefer it for it's flexibility, but I didn't take this road because I am pretty sure the engine has some race when you have many merge rules. Currently people have no chance to run into them. But if they start switching labels to change the merge rule, this will increase a lot the change to trigger them.

That's said, while I'm writing this, I found a simple idea to fix up the race I have in mind.

I will split this PR to merge the first part and propose the priority thing into another one.

Reviewer: Awesome <3
